test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,461 @@
+  parity_test_error_msg);
+""")
+
+CHECK_MODULE_ATTR_EQUALITY = Template("""\
+TORCH_CHECK(
+  check_ivalue_equality(
+    ${script_module_prefix}.get_attribute("${attr_name}"), c10::IValue(${cpp_module_prefix}->${attr_name})),
+  parity_test_error_msg);
+""")
+
+TORCH_NN_MODULE_TEST_CTOR_ARGS = Template("""\n
+void ${module_name}_test_ctor_args() {
+  ${module_qualified_name} m_init_by_cpp(${module_option});
+}
+""")
+
+TORCH_NN_MODULE_TEST_INIT = Template("""\n
+void ${module_variant_name}_test_init(
+    const std::string& saved_module_path,
+    const std::string& device) {
+  torch::jit::script::Module m_init_by_python = torch::jit::load(saved_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} m_init_by_cpp${cpp_constructor_args};
+  m_init_by_cpp->to(device);
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_TEST_FORWARD = Template("""\n
+void ${module_variant_name}_test_forward(
+    const std::string& saved_module_path,
+    const std::string& device,
+    torch::Tensor python_output,
+    ${input_arg_declarations}) {
+  torch::manual_seed(2);
+  ${module_qualified_name} module${cpp_constructor_args};
+  torch::load(module, saved_module_path);
+  module->to(device);
+
+  auto cpp_output = module(${input_args});
+
+  TORCH_CHECK(
+    cpp_output.sizes().vec() == python_output.sizes().vec() &&
+    cpp_output.allclose(python_output),
+    parity_test_error_msg, ": forward output doesn't match");
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_TEST_BACKWARD = Template("""\n
+void ${module_variant_name}_test_backward(
+    const std::string& saved_module_path,
+    const std::string& saved_grad_module_path,
+    const std::string& device,
+    ${input_arg_declarations}) {
+  ${module_qualified_name} grad_module${cpp_constructor_args};
+  torch::load(grad_module, saved_grad_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} module${cpp_constructor_args};
+  torch::load(module, saved_module_path);
+  module->to(device);
+
+  auto cpp_output = module(${input_args});
+  cpp_output.sum().backward();
+
+  for (size_t i = 0; i < module->parameters().size(); i++) {
+    auto named_param = module->named_parameters()[i];
+    auto grad = grad_module->parameters()[i];
+    TORCH_CHECK(
+      named_param->grad().allclose(grad),
+      parity_test_error_msg, ": ", "gradient value of `", named_param.key(), "` doesn't match");
+  }
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_IGNORED_ATTRS = {
+    '_backend', '_parameters', '_buffers', '_backward_hooks', '_forward_hooks', '_forward_pre_hooks',
+    '_state_dict_hooks', '_load_state_dict_pre_hooks', '_modules', 'training', 'has_parity',
+}
+
+class TestCppApiParity(common.TestCase):
+    def _python_arg_to_cpp_arg(self, python_arg):
+        if type(python_arg) == int:
+            return 'int64_t', str(python_arg)
+        elif type(python_arg) == float:
+            return 'double', str(python_arg)
```

Comment:
I'm not sure that `str(float('NaN'))` (which evaluates to `'nan'`) is valid c++. However, I can't think of a module yet that takes in NaN so we can worry about that later